ibis/backends/pandas/executor/generic.py

```diff
@@ -0,0 +1,319 @@
+
+_binary_operations = {
+    ops.Greater: operator.gt,
+    ops.Less: operator.lt,
+    ops.LessEqual: operator.le,
+    ops.GreaterEqual: operator.ge,
+    ops.Equals: operator.eq,
+    ops.NotEquals: operator.ne,
+    ops.And: operator.and_,
+    ops.Or: operator.or_,
+    ops.Xor: operator.xor,
+    ops.Add: operator.add,
+    ops.Subtract: operator.sub,
+    ops.Multiply: operator.mul,
+    ops.Divide: operator.truediv,
+    ops.FloorDivide: operator.floordiv,
+    ops.Modulus: operator.mod,
+    ops.Power: operator.pow,
+    ops.IdenticalTo: lambda x, y: (x == y) | (pd.isnull(x) & pd.isnull(y)),
+    ops.BitwiseXor: lambda x, y: np.bitwise_xor(x, y),
+    ops.BitwiseOr: lambda x, y: np.bitwise_or(x, y),
+    ops.BitwiseAnd: lambda x, y: np.bitwise_and(x, y),
+    ops.BitwiseLeftShift: lambda x, y: np.left_shift(x, y).astype("int64"),
+    ops.BitwiseRightShift: lambda x, y: np.right_shift(x, y).astype("int64"),
+    ops.Atan2: np.arctan2,
+}
+
+
+@execute.register(ops.Unary)
+def execute_unary(op, arg):
+    return _unary_operations[type(op)](arg)
+
+
+@execute.register(ops.Binary)
+def execute_equals(op, left, right):
+    return _binary_operations[type(op)](left, right)
+
+
+def mapdecimal(func, s):
+    def wrapper(x):
+        try:
+            return func(x)
+        except decimal.InvalidOperation:
+            return decimal.Decimal("NaN")
+
+    return s.map(wrapper)
+
+
+@execute.register(ops.Sqrt)
+def execute_sqrt(op, arg):
+    if op.arg.dtype.is_decimal():
+        return mapdecimal(lambda x: x.sqrt(), arg)
+    else:
+        return np.sqrt(arg)
+
+
+@execute.register(ops.Ln)
+def execute_ln(op, arg):
+    if op.arg.dtype.is_decimal():
+        return mapdecimal(lambda x: x.ln(), arg)
+    else:
+        return np.log(arg)
+
+
+@execute.register(ops.Log2)
+def execute_log2(op, arg):
+    if op.arg.dtype.is_decimal():
+        # TODO(kszucs): this doesn't support columnar shaped base
+        baseln = decimal.Decimal(2).ln()
+        return mapdecimal(lambda x: x.ln() / baseln, arg)
+    else:
+        return np.log2(arg)
+
+
+@execute.register(ops.Log10)
+def execute_log10(op, arg):
+    if op.arg.dtype.is_decimal():
+        return mapdecimal(lambda x: x.log10(), arg)
+    else:
+        return np.log10(arg)
+
+
+@execute.register(ops.Log)
+def execute_log(op, arg, base):
+    if op.arg.dtype.is_decimal():
+        # TODO(kszucs): this doesn't support columnar shaped base
+        baseln = decimal.Decimal(base).ln()
+        return mapdecimal(lambda x: x.ln() / baseln, arg)
+    elif base is None:
+        return np.log(arg)
+    else:
+        return np.log(arg) / np.log(base)
+
+
+@execute.register(ops.Exp)
+def execute_exp(op, arg):
+    if op.arg.dtype.is_decimal():
+        return mapdecimal(lambda x: x.exp(), arg)
+    else:
+        return np.exp(arg)
+
+
+@execute.register(ops.Round)
+def execute_round(op, arg, digits):
+    if op.arg.dtype.is_decimal():
+        if digits is None:
+            return arg.map(round)
```

Comment:
Anywhere you're using `Series.map` should probably be called with `na_action="ignore"`. It'll simplify the code.

Reply:
Updated.